### PR TITLE
All the tickets API and functionality

### PR DESCRIPTION
**Pull request by MarcAntoineHebert**

### CODE REVIEW

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.ArrayList;
+
+/**
+ * Created by Marc-Antoine on 17-06-19.
+ */
+
+@RestController
+@RequestMapping("/api")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Should be @RequestMapping(value = "/api" + V1 + "/tickets")

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+public class TicketAPI {
+    private IUserService userService;
+
+    @Autowired
+    UserRepository userRepository;
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Should use the userservice instead

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    @Autowired
+    public TicketAPI(IUserService userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping("/users/{email}/tickets")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Mapping that starts with /users should be in the UserApi

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    public String addTicket(@ModelAttribute Ticket ticket, @PathVariable(value="email") String email){
+        User user = userRepository.findByUsername(email);
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Should use the service for that

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        //Add the ticket in the user tickets list
+        ArrayList<Ticket> tickets= user.getTickets();
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: What if the user is null ?

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        tickets.add(ticket);
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: User should have a addTicket method, that way we do not depend on the tickets datastructure implementation

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        user.setTickets(tickets);
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Unprimitive object are passed by reference, no need for this

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+        //Commit the changes in the DB
+        userRepository.save(user);
+
+        return "The ticket has been added";
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Should return a response entity

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @GetMapping("/users/{email}/tickets")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: UserApi mapping

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    public ArrayList<Ticket> getUserTickets(@PathVariable(value="email") String email){
+        User user = userRepository.findByUsername(email);
+        ArrayList<Ticket> tickets = user.getTickets();
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Unused variable, could directly return the object, but we should return a response entity with the GetResponse as entity and http status 200

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        return tickets;
+    }
+
+    @GetMapping("/users/friends/{email}/tickets")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: User mapping

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    public ArrayList<Ticket> getFriendTickets(@PathVariable(value="email") String email){
+        User user = userRepository.findByUsername(email);
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Use the user service instead

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        ArrayList<Ticket> tickets = user.getTickets();
+        return tickets;
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Same as :point_up:  should return a ResponseEntity with GetReponse

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @PostMapping("/ticket/transfer")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: ticket should be tickets and added in the main requestmapping (see first comment) that mapping should only be /transfer

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    public String transferTicket(@ModelAttribute Ticket receivedTicket, @ModelAttribute User receiver, Authentication authentication){
+        try {
+            User receiverUser = userRepository.findByUsername(receiver.getUsername());
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Use the service

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            User providerUser = userService.findById(authentication.getName());
+
+            //Retrieve tickets from both users
+            ArrayList<Ticket> receiverTickets = receiverUser.getTickets();
+            ArrayList<Ticket> providerTickets = providerUser.getTickets();
+
+            //Walkaround to retrieve the ticket with the ID
+            Ticket ticket = new Ticket();
+            for (Ticket ticketFor:providerTickets) {
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: userService.getTicketFrom(user, ticket) ?

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+                if(ticketFor.getId().equals(receivedTicket.getId()))
+                    ticket = ticketFor;
+            }
+
+            //Add the ticket in the receiver tickets list
+            receiverTickets.add(ticket);
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Should not work with the list directly and should use the service isntead

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+            //Remove the ticket from the provider tickets list
+            providerTickets.remove(ticket);
+            providerUser.setTickets(providerTickets);
+
+            //Commit the changes in the DB
+            userRepository.save(receiverUser);
+            userRepository.save(providerUser);
+
+            return ("The ticket has been transferred to: " + receiverUser.getUsername());
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Return response entity

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        catch (ResourceNotFoundException e) {
+            return ("Error ! The ticket cannot be transferred");
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: http 400 ? 500 ?

File: src/main/java/com/arachidsocial/controllers/pages/MyProfilePage.java [hunk 20]
```diff
@@ -38,6 +39,9 @@ public String loadpage(Model model, Authentication authentication) {
         model.addAttribute("friends", friends);
         model.addAttribute("friendRequests", friendRequests);
 
+        ArrayList<Ticket> tickets = loggedUser.getTickets();
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/pages/MyProfilePage.java]: No need for this variable

File: src/main/java/com/arachidsocial/controllers/pages/PublicProfilePage.java [hunk 21]
```diff
@@ -44,6 +45,9 @@ public String loadpage(Model model, Authentication authentication, @PathVariable
         model.addAttribute("user", user);
         model.addAttribute("userFriendRequests", userFriendRequests);
 
+        ArrayList<Ticket> tickets = user.getTickets();
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/pages/PublicProfilePage.java]: Same :point_up:

File: src/main/java/com/arachidsocial/controllers/pages/TicketDetailsPage.java [hunk 22]
```diff
@@ -17,8 +19,8 @@
     @Autowired
     UserRepository userRepository;
 
-    @GetMapping("/ticket-info")
-    public String loadpage(Model model, Authentication authentication) {
+    @GetMapping("/ticket-info/{ticketId}/")
```
Comment (banctilrobitaille) [inline src/main/java/com/arachidsocial/controllers/pages/TicketDetailsPage.java]: /tickets/{id}/details  would be cleaner

File: src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java (hunk 9 above)
Comment (MarcAntoineHebert) [inline src/main/java/com/arachidsocial/controllers/api/v1/resources/TicketAPI.java]: Tu veux quoi comme URL pour le GET ?

**Pull request merged**